Mobile : Fixes white screen after deleting last notebook and solves issue  #2727

ishantgupta777 opened the pull request:
closes #2727 
Platform: Mobile

I fixed the issue of white blank screen after deleting the last notebook. Please have a look. I redirected the user to the All_Notes page where they can add a new notebook.
@laurent22  is this good?

ishantgupta777 commented:
@PackElend  can you please add a gsoc-2020 label in this PR.

ishantgupta777 commented:
@laurent22  can you please check this PR. I have removed extra lines from this PR. Tell me, If I need to change anything else in this PR.

PackElend commented:
> @laurent22 can you please check this PR.

please minimize mentioning those folks, all PRs will be watched, in particular those labeled with gsoc-2020. Please stick to the rules in the GSoC live blog in the forum

ishantgupta777 commented:
> > @laurent22 can you please check this PR.
> 
> please minimize mentioning those folks, all PRs will be watched, in particular those labeled with gsoc-2020. Please stick to the rules in the GSoC live blog in the forum

Oh okay okay, I am really sorry for the problem caused. Just to be sure, I just have to tag you to label my PR, that's it, right?

PackElend commented:
no worries you can label me and bedwardly-down but if all of you guys mention the coders in your PR and comments the going to ignore them and they are loosing control of their inbox

ishantgupta777 commented:
> no worries you can label me and bedwardly-down but if all of you guys mention the coders in your PR and comments the going to ignore them and they are loosing control of their inbox

Oh okay okay, got it, thanks. I will keep this in mind.

laurent22 commented on ReactNativeClient/lib/components/side-menu-content.js:
```diff
@@ -148,8 +148,10 @@ class SideMenuContentComponent extends Component {
+								onPress: async () => {
+									await Folder.delete(folder.id);
+									if (this.props.folders.length==0)
+										this.allNotesButton_press();
```
What if the last notebook is deleted via synchronisation? Also I believe this happens not just when the last notebook is deleted but when the currently opened notebook is deleted (eg. via sync). This is a tricky problem that needs more work.

ishantgupta777 replied:
Oh, I got your point, sir, I didn't check it with sync. Let me do the manual testing with sync and I will let you know. I tested it on mobile, this is the case only when the last notebook was getting deleted, not any notebook. It might be the case with sync. Can you suggest what might the approach to fix this, or should I try on my own first?

ishantgupta777 replied:
As you suspected, I checked with sync and its not working with sync. But this is only the case with last notebook, all the notebooks when gets deleted,  the page of next notebook gets open, even in sync. So I will try to fix sync part of this also. I am trying to understand sync part of code right now and will get back to you as soon as I fix this.

ishantgupta777 replied:
`if (!this.cancelling()) {
					for (let i = 0; i < localFoldersToDelete.length; i++) {
						const item = localFoldersToDelete[i];
						const noteIds = await Folder.noteIds(item.id);
						if (noteIds.length) {
							// CONFLICT
							await Folder.markNotesAsConflict(item.id);
						}
						await Folder.delete(item.id, { deleteChildren: false, changeSource: ItemChange.SOURCE_SYNC, trackDeleted: false });
					}
				}`

These are the code in syncronizer.js lines starting from 787. 
Sir, if I dispatch an action here, which will check if all the folders has been deleted, then it will redirect to home page. I think this should work as all the other test cases are getting passed. What do you think? And sir, I don't know why, sync is not working in emulators for me, but working in released apps and desktop clients. So, there is no way for me to check sync after changing something in my code. Can you please suggest something.

mic704b replied:
@ishantgupta777 please take note of the rules for PR titles. Generally they should start with "Desktop:" or "Mobile:".  Please fix your titles.

@tessus could we update the instructions to include extra prefixes like All, Test, Docs?

mic704b commented:
@ishantgupta777 please take note of the rules for PR titles. Generally they should start with "Desktop:" or "Mobile:".  Please fix your titles.

@tessus could we update the instructions to include extra prefixes like All, Test, Docs?

tessus commented:
@mic704b we should ask Laurent. I'd be all for it, but maybe there's a reason for the current system. Maybe the script that generates the changelog has something to do with it.

The current PR template states:

_If it's not related to any platform (such as a translation, change to the documentation, etc.), simply don't add a platform._

But I usually add `Translation: ` or `Documentation: ` to PRs I merge. Or call the branches `infra/...` when it's a change related to the projects infrastructure (scripts, tests, ...).

Laurent hasn't said anything so far (and I've been adding these prefixes for a while now), thus I think adding the prefixes you mentioned should be fine.

I think it is a good idea to add more prefixes. It's too bad that git doesn't use tags (as in labels), this would be so much more convenient.

mic704b commented:
> but maybe there's a reason for the current system. Maybe the script that generates the changelog has something to do with it.

Thanks, that didn't even cross my mind.

laurent22 commented:
I'm not very keen on the current implementation. It adds business logic in the view, while that should most likely be in the reducer. Please see if you can find a fix without changing the view @ishantgupta777 and if it's too complicated let's close the pull request.

ishantgupta777 commented:
> I'm not very keen on the current implementation. It adds business logic in the view, while that should most likely be in the reducer. Please see if you can find a fix without changing the view @ishantgupta777 and if it's too complicated let's close the pull request.

Sir, can you please check now, as you suggested I moved the logic from view to actions. I added a new action to handle the deletion of last folder and directly dispatched it from the Folder class, as while syncing and without it, folders are getting deleted from the delete method of Folder class, there shouldn't be any problem while syncing. Can you please check this PR. Thanks.

laurent22 commented:
You copied and pasted the code from NAV_GO. In this project we try as much as possible to avoid code duplication as it is a maintenance burden.

ishantgupta777 commented:
> You copied and pasted the code from NAV_GO. In this project we try as much as possible to avoid code duplication as it is a maintenance burden.

I didn't just copied the code, i copied it and removed the extra part which is not needed. I am trying to use code and functions already written, as to save time and resources. But if you want I can write that code by myself, that won't be a much issue, but you can see I modified it and removed the extra part. I didn't copy any of my code from anywhere else and copied from nav_go because I think that will ensure I don't miss any imp edge case.

ishantgupta777 commented:
> > You copied and pasted the code from NAV_GO. In this project we try as much as possible to avoid code duplication as it is a maintenance burden.
> 
> I didn't just copied the code, i copied it and removed the extra part which is not needed. I am trying to use code and functions already written, as to save time and resources. But if you want I can write that code by myself, that won't be a much issue, but you can see I modified it and removed the extra part. I didn't copy any of my code from anywhere else and copied from nav_go because I think that will ensure I don't miss any imp edge case.

laurent22 commented:
Sorry I feel this PR is too complex to be acceptable. The current situation is not great, but the user can just press on the menu button to add a new notebook.
